crates/pipeline_manager/src/local_runner.rs

```diff
@@ -0,0 +1,773 @@
+            })?;
+
+        // Locate project executable.
+        let executable_ref = self
+            .db
+            .lock()
+            .await
+            .get_compiled_binary_ref(program_id, pr.program.version)
+            .await?;
+        let fetched_executable = binary_ref_to_url(&executable_ref, pipeline_id).await?;
+
+        // Run executable, set current directory to pipeline directory, pass metadata
+        // file and config as arguments.
+        let pipeline_process = Command::new(fetched_executable)
+            .current_dir(self.config.pipeline_dir(pipeline_id))
+            .arg("--config-file")
+            .arg(&config_file_path)
+            .arg("--metadata-file")
+            .arg(&metadata_file_path)
+            .stdin(Stdio::null())
+            .spawn()
+            .map_err(|e| RunnerError::PipelineStartupError {
+                pipeline_id,
+                error: e.to_string(),
+            })?;
+
+        Ok(PipelineHandle { pipeline_process })
+    }
+
+    /// Parse `ErrorResponse` from JSON. On error, builds an `ErrorResponse`
+    /// with the originaln JSON content.
+    fn error_response_from_json(
+        pipeline_id: PipelineId,
+        status: StatusCode,
+        json: &JsonValue,
+    ) -> ErrorResponse {
+        ErrorResponse::deserialize(json).unwrap_or_else(|_| {
+            ErrorResponse::from(&RunnerError::HttpForwardError {
+                pipeline_id,
+                error: format!("Pipeline returned HTTP status {status}, response body:{json:#}"),
+            })
+        })
+    }
+}
+
+/// Starts a runner that executes pipelines locally
+///
+/// # Starting a pipeline
+///
+/// Starting a pipeline amounts to running the compiled executable with
+/// selected config, and monitoring the pipeline log file for either
+/// "Started HTTP server on port XXXXX" or "Failed to create server
+/// [detailed error message]".  In the former case, the port number is
+/// recorded in the database.  In the latter case, the error message is
+/// returned to the client.
+///
+/// # Shutting down a pipeline
+///
+/// To shutdown the pipeline, the runner sends a `/shutdown` HTTP request to the
+/// pipeline.  This request is asynchronous: the pipeline may continue running
+/// for a few seconds after the request succeeds.
+pub async fn run(db: Arc<Mutex<ProjectDB>>, config: &LocalRunnerConfig) {
+    let runner_task = spawn(reconcile(db, Arc::new(config.clone())));
+    runner_task.await.unwrap().unwrap();
+}
+
+async fn reconcile(
+    db: Arc<Mutex<ProjectDB>>,
+    config: Arc<LocalRunnerConfig>,
+) -> Result<(), ManagerError> {
+    let pipelines: Mutex<BTreeMap<PipelineId, Arc<Notify>>> = Mutex::new(BTreeMap::new());
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+    tokio::spawn(crate::db_notifier::listen(db.clone(), tx));
+    loop {
+        trace!("Waiting for notification");
+        if let Some(DbNotification::Pipeline(op, tenant_id, pipeline_id)) = rx.recv().await {
```

Comment:
We should also stop and remove the pipeline automaton when the pipeline gets deleted (this was already missing in my implementation).

Reply:
Ack. I'll remove the automata on pipeline deletes.